src/main/java/org/jenkinsci/plugins/buildnameupdater/BuildNameUpdater.java

```diff
@@ -0,0 +1,243 @@
+import hudson.model.AbstractProject;
+import hudson.model.BuildListener;
+import hudson.model.FreeStyleProject;
+import hudson.remoting.VirtualChannel;
+import hudson.tasks.BuildStepDescriptor;
+import hudson.tasks.Builder;
+import hudson.util.FormValidation;
+import org.jenkinsci.plugins.tokenmacro.MacroEvaluationException;
+import org.jenkinsci.plugins.tokenmacro.TokenMacro;
+import org.kohsuke.stapler.DataBoundConstructor;
+import org.kohsuke.stapler.QueryParameter;
+
+import javax.servlet.ServletException;
+import java.io.BufferedReader;
+import java.io.File;
+import java.io.FileReader;
+import java.io.IOException;
+
+/**
+ *
+ * This plugin replace the build name with the first line from a file on a slave.
+ *
+ * @author Lev Mishin
+ */
+public class BuildNameUpdater extends Builder {
+
+    private final String buildName;
+    private final String macroTemplate;
+    private final boolean fromFile;
+    private final boolean fromMacro;
+    private final boolean macroFirst;
+
+    // Fields in config.jelly must match the parameter names in the "DataBoundConstructor"
+    @DataBoundConstructor
+    public BuildNameUpdater(BuildNameFromFile nameFromFile, BuildNameFromMacro nameFromMacro) {
+        if (nameFromFile != null) {
+            this.buildName = nameFromFile.buildName;
+            this.fromFile = true;
+        }
+        else {
+            this.buildName = "";
+            this.fromFile = false;
+        }
+
+        if (nameFromMacro != null) {
+            this.macroTemplate = nameFromMacro.macroTemplate;
+            this.fromMacro = true;
+            this.macroFirst = nameFromMacro.macroFirst;
+        }
+        else {
+            this.macroTemplate = "";
+            this.fromMacro = false;
+            this.macroFirst = false;
+        }
+    }
+
+    @SuppressWarnings("unused")
+    public boolean getFromFile() {
+        return fromFile;
+    }
+
+    @SuppressWarnings("unused")
+    public boolean getMacroFirst() {
+        return macroFirst;
+    }
+
+    @SuppressWarnings("unused")
+    public boolean getFromMacro() {
+        return fromMacro;
+    }
+
+    @SuppressWarnings("unused")
+    public String getBuildName() {
+        return buildName;
+    }
+
+    @SuppressWarnings("unused")
+    public String getMacroTemplate() {
+        return macroTemplate;
+    }
+
+    @Override
+    public boolean perform(AbstractBuild build, Launcher launcher, BuildListener listener) {
+        String buildNameToSet = "";
+
+        if (fromFile){
+            buildNameToSet = readFromFile(build, listener, buildName);
+        }
+
+        if(fromMacro){
+            String evaluatedMacro = getFromMacro(build, listener, macroTemplate);
+
+            listener.getLogger().println("Evaluated macro: '" + evaluatedMacro + "'");
+
+            buildNameToSet = macroFirst ? evaluatedMacro + buildNameToSet : buildNameToSet + evaluatedMacro;
+        }
+
+        if (buildNameToSet != null && !buildNameToSet.isEmpty()) {
```

Comment:
done, now I'm using StringUtils.IsNotBlank() in order to check that string is not null or empty